tools/sourcemap-tools/src/SourceMapUploader.ts

```diff
@@ -0,0 +1,151 @@
+import crypto from 'crypto';
+import fs from 'fs';
+import http from 'http';
+import https from 'https';
+import { Readable } from 'stream';
+import { SOURCEMAP_DEBUG_ID_KEY } from './DebugIdGenerator';
+
+export const DEBUG_ID_QUERY = 'symbolication_id';
+
+interface Sourcemap {
+    version: number;
+    [SOURCEMAP_DEBUG_ID_KEY]?: string;
+}
+
+interface UploadResponse {
+    response: 'ok' | string;
+    _rxid: string;
+}
+
+export interface UploadResult {
+    rxid: string;
+    debugId: string;
+}
+
+export class SourceMapUploader {
+    private readonly _url: URL;
+
+    constructor(url: string | URL) {
+        this._url = new URL(url);
+    }
+
+    public async upload(fileStream: Readable, debugId?: string): Promise<UploadResult>;
+    public async upload(filePath: string, debugId?: string): Promise<UploadResult>;
+    public async upload(pathOrStream: string | Readable, debugId?: string): Promise<UploadResult> {
+        if (typeof pathOrStream === 'string') {
+            pathOrStream = fs.createReadStream(pathOrStream);
+        }
+
+        const content = await this.readStreamToEnd(pathOrStream);
+        const sourcemap = JSON.parse(content.toString('utf-8'));
+
+        pathOrStream.destroy();
+
+        return this.uploadSourcemap(sourcemap, debugId);
+    }
+
+    public async uploadContent(content: string, debugId?: string): Promise<UploadResult>;
+    public async uploadContent(content: object, debugId?: string): Promise<UploadResult>;
+    public uploadContent(content: string | object, debugId?: string): Promise<UploadResult> {
+        if (typeof content === 'string') {
+            content = JSON.parse(content) as object;
+        }
+
+        return this.uploadSourcemap(content, debugId);
+    }
+
+    private async uploadSourcemap(sourcemap: unknown, debugId?: string): Promise<UploadResult> {
+        this.assertValidSourcemap(sourcemap);
+
+        if (!debugId) {
+            debugId = sourcemap[SOURCEMAP_DEBUG_ID_KEY];
+
+            if (!debugId) {
+                debugId = crypto.randomUUID();
+                console.warn(`Sourcemap does not have a debug ID. Using ${debugId}`);
+            }
+        }
+
+        const uploadUrl = this.buildUploadUrl(debugId);
+        const protocol = uploadUrl.protocol === 'https:' ? https : http;
+
+        return new Promise<UploadResult>((resolve, reject) => {
+            const request = protocol.request(
+                {
```

Comment:
you can also use just url as a fist parameter, You don't need to pass all URL parameters into options.

Reply:
I cannot, because I need to set the `method` and `rejectUnauthorized`.

Reply:
method and rejectUnauthorized can be passed as a second parameter

Reply:
done